File: clients/messagedb_ex/test/encoding_test.exs
defmodule MessagedbEx.EncodingTest do
  use ExUnit.Case, async: true
  import MessagedbEx.TestHelper

  setup do
    {client, namespace_id, _token} = create_test_namespace("encoding")

    on_exit(fn ->
      cleanup_namespace(client, namespace_id)
    end)

    %{client: client}
  end

  test "ENCODING-001: UTF-8 text in data", %{client: client} do
    stream = unique_stream()
    utf8_text = "Hello 世界 🌍 émojis"
    message = %{type: "TestEvent", data: %{text: utf8_text}}

    {:ok, _result, client} = MessagedbEx.stream_write(client, stream, message)
    {:ok, [msg], _client} = MessagedbEx.stream_get(client, stream)

    [_id, _type, _pos, _gpos, data, _metadata, _time] = msg
    assert data["text"] == utf8_text
  end

  test "ENCODING-002: Unicode in metadata", %{client: client} do
    stream = unique_stream()
    utf8_text = "Test 测试 🎉"
    message = %{type: "TestEvent", data: %{}, metadata: %{description: utf8_text}}

    {:ok, _result, client} = MessagedbEx.stream_write(client, stream, message)
    {:ok, [msg], _client} = MessagedbEx.stream_get(client, stream)

    [_id, _type, _pos, _gpos, _data, metadata, _time] = msg
    assert metadata["description"] == utf8_text
  end

  test "ENCODING-003: Special characters in stream name", %{client: client} do
    stream = "test-stream_123.abc-#{unique_suffix()}"
    message = %{type: "TestEvent", data: %{}}

    # Should succeed or give clear error
    result = MessagedbEx.stream_write(client, stream, message)
    assert match?({:ok, _, _}, result) or match?({:error, _}, result)
  end

  test "ENCODING-004: Empty string values", %{client: client} do
    stream = unique_stream()
    message = %{type: "TestEvent", data: %{empty_string: ""}}

    {:ok, _result, client} = MessagedbEx.stream_write(client, stream, message)
    {:ok, [msg], _client} = MessagedbEx.stream_get(client, stream)

    [_id, _type, _pos, _gpos, data, _metadata, _time] = msg
    assert data["empty_string"] == ""
    assert data["empty_string"] != nil
  end

  test "ENCODING-005: Boolean values", %{client: client} do
    stream = unique_stream()
    message = %{type: "TestEvent", data: %{is_true: true, is_false: false}}

    {:ok, _result, client} = MessagedbEx.stream_write(client, stream, message)
    {:ok, [msg], _client} = MessagedbEx.stream_get(client, stream)

    [_id, _type, _pos, _gpos, data, _metadata, _time] = msg
    assert data["is_true"] == true
    assert data["is_false"] == false
    assert is_boolean(data["is_true"])
    assert is_boolean(data["is_false"])
  end

  test "ENCODING-006: Null values", %{client: client} do
    stream = unique_stream()
    message = %{type: "TestEvent", data: %{null_value: nil}}

    {:ok, _result, client} = MessagedbEx.stream_write(client, stream, message)
    {:ok, [msg], _client} = MessagedbEx.stream_get(client, stream)

    [_id, _type, _pos, _gpos, data, _metadata, _time] = msg
    assert data["null_value"] == nil
    assert Map.has_key?(data, "null_value")
  end

  test "ENCODING-007: Numeric values", %{client: client} do
    stream = unique_stream()

    message = %{
      type: "TestEvent",
      data: %{
        integer: 42,
        float: 3.14159,
        negative: -100,
        zero: 0
      }
    }

    {:ok, _result, client} = MessagedbEx.stream_write(client, stream, message)
    {:ok, [msg], _client} = MessagedbEx.stream_get(client, stream)

    [_id, _type, _pos, _gpos, data, _metadata, _time] = msg
    assert data["integer"] == 42
    assert_in_delta data["float"], 3.14159, 0.00001
    assert data["negative"] == -100
    assert data["zero"] == 0
  end

  test "ENCODING-008: Nested objects", %{client: client} do
    stream = unique_stream()

    nested = %{
      level1: %{
        level2: %{
          level3: %{
            value: "deep"
          }
        }
      }
    }

    message = %{type: "TestEvent", data: nested}

    {:ok, _result, client} = MessagedbEx.stream_write(client, stream, message)
    {:ok, [msg], _client} = MessagedbEx.stream_get(client, stream)

    [_id, _type, _pos, _gpos, data, _metadata, _time] = msg
    assert data["level1"]["level2"]["level3"]["value"] == "deep"
  end

  test "ENCODING-009: Arrays in data", %{client: client} do
    stream = unique_stream()

    message = %{
      type: "TestEvent",
      data: %{
        items: [1, "two", %{three: 3}, nil, true]
      }
    }

    {:ok, _result, client} = MessagedbEx.stream_write(client, stream, message)
    {:ok, [msg], _client} = MessagedbEx.stream_get(client, stream)

    [_id, _type, _pos, _gpos, data, _metadata, _time] = msg
    items = data["items"]
    assert Enum.at(items, 0) == 1
    assert Enum.at(items, 1) == "two"
    assert Enum.at(items, 2) == %{"three" => 3}
    assert Enum.at(items, 3) == nil
    assert Enum.at(items, 4) == true
  end

  test "ENCODING-010: Large message payload", %{client: client} do
    stream = unique_stream()

    # Create a large data object (~10KB)
    large_data = %{
      items: Enum.map(1..1000, fn i -> %{id: i, name: "Item #{i}", value: i * 10} end)
    }

    message = %{type: "TestEvent", data: large_data}

    # Should succeed (under 1MB limit)
    assert {:ok, _result, client} = MessagedbEx.stream_write(client, stream, message)

    # Verify it reads back correctly
    {:ok, [msg], _client} = MessagedbEx.stream_get(client, stream)
    [_id, _type, _pos, _gpos, data, _metadata, _time] = msg
    assert length(data["items"]) == 1000
  end
end

File: clients/eventodb-node/tests/encoding.test.ts
import { describe, test, expect, afterEach } from 'vitest';
import { setupTest, randomStreamName, type TestContext } from './helpers.js';

describe('ENCODING Tests', () => {
  const contexts: TestContext[] = [];

  afterEach(async () => {
    await Promise.all(contexts.map(ctx => ctx.cleanup()));
    contexts.length = 0;
  });

  test('ENCODING-001: UTF-8 text in data', async () => {
    const ctx = await setupTest('encoding-001');
    contexts.push(ctx);

    const stream = randomStreamName();
    const utf8Text = 'Hello 世界 🌍 émojis';
    
    await ctx.client.streamWrite(stream, {
      type: 'TestEvent',
      data: { text: utf8Text }
    });

    const messages = await ctx.client.streamGet(stream);
    const [, , , , data] = messages[0];
    expect(data.text).toBe(utf8Text);
  });

  test('ENCODING-002: Unicode in metadata', async () => {
    const ctx = await setupTest('encoding-002');
    contexts.push(ctx);

    const stream = randomStreamName();
    const unicodeText = 'Test 测试 🎉';
    
    await ctx.client.streamWrite(stream, {
      type: 'TestEvent',
      data: { test: true },
      metadata: { description: unicodeText }
    });

    const messages = await ctx.client.streamGet(stream);
    const [, , , , , metadata] = messages[0];
    expect(metadata?.description).toBe(unicodeText);
  });

  test('ENCODING-003: Special characters in stream name', async () => {
    const ctx = await setupTest('encoding-003');
    contexts.push(ctx);

    const stream = 'test-stream_123.abc';
    
    // Should either succeed or give clear error
    const result = await ctx.client.streamWrite(stream, {
      type: 'TestEvent',
      data: { test: true }
    });

    expect(result.position).toBeGreaterThanOrEqual(0);
  });

  test('ENCODING-004: Empty string values', async () => {
    const ctx = await setupTest('encoding-004');
    contexts.push(ctx);

    const stream = randomStreamName();
    
    await ctx.client.streamWrite(stream, {
      type: 'TestEvent',
      data: { emptyString: '' }
    });

    const messages = await ctx.client.streamGet(stream);
    const [, , , , data] = messages[0];
    expect(data.emptyString).toBe('');
    expect(data.emptyString).not.toBe(null);
  });

  test('ENCODING-005: Boolean values', async () => {
    const ctx = await setupTest('encoding-005');
    contexts.push(ctx);

    const stream = randomStreamName();
    
    await ctx.client.streamWrite(stream, {
      type: 'TestEvent',
      data: { isTrue: true, isFalse: false }
    });

    const messages = await ctx.client.streamGet(stream);
    const [, , , , data] = messages[0];
    expect(data.isTrue).toBe(true);
    expect(data.isFalse).toBe(false);
    expect(typeof data.isTrue).toBe('boolean');
    expect(typeof data.isFalse).toBe('boolean');
  });

  test('ENCODING-006: Null values', async () => {
    const ctx = await setupTest('encoding-006');
    contexts.push(ctx);

    const stream = randomStreamName();
    
    await ctx.client.streamWrite(stream, {
      type: 'TestEvent',
      data: { nullValue: null }
    });

    const messages = await ctx.client.streamGet(stream);
    const [, , , , data] = messages[0];
    expect(data.nullValue).toBe(null);
    expect(data.nullValue).not.toBe(undefined);
  });

  test('ENCODING-007: Numeric values', async () => {
    const ctx = await setupTest('encoding-007');
    contexts.push(ctx);

    const stream = randomStreamName();
    
    await ctx.client.streamWrite(stream, {
      type: 'TestEvent',
      data: {
        integer: 42,
        float: 3.14159,
        negative: -100,
        zero: 0
      }
    });

    const messages = await ctx.client.streamGet(stream);
    const [, , , , data] = messages[0];
    expect(data.integer).toBe(42);
    expect(data.float).toBe(3.14159);
    expect(data.negative).toBe(-100);
    expect(data.zero).toBe(0);
    expect(typeof data.integer).toBe('number');
    expect(typeof data.float).toBe('number');
  });

  test('ENCODING-008: Nested objects', async () => {
    const ctx = await setupTest('encoding-008');
    contexts.push(ctx);

    const stream = randomStreamName();
    const nestedData = {
      level1: {
        level2: {
          level3: {
            value: 'deep'
          }
        }
      }
    };
    
    await ctx.client.streamWrite(stream, {
      type: 'TestEvent',
      data: nestedData
    });

    const messages = await ctx.client.streamGet(stream);
    const [, , , , data] = messages[0];
    expect(data).toEqual(nestedData);
    expect(data.level1.level2.level3.value).toBe('deep');
  });

  test('ENCODING-009: Arrays in data', async () => {
    const ctx = await setupTest('encoding-009');
    contexts.push(ctx);

    const stream = randomStreamName();
    const arrayData = {
      items: [1, 'two', { three: 3 }, null, true]
    };
    
    await ctx.client.streamWrite(stream, {
      type: 'TestEvent',
      data: arrayData
    });

    const messages = await ctx.client.streamGet(stream);
    const [, , , , data] = messages[0];
    expect(data).toEqual(arrayData);
    expect(Array.isArray(data.items)).toBe(true);
    expect(data.items).toHaveLength(5);
    expect(data.items[0]).toBe(1);
    expect(data.items[1]).toBe('two');
    expect(data.items[2]).toEqual({ three: 3 });
    expect(data.items[3]).toBe(null);
    expect(data.items[4]).toBe(true);
  });

  test('ENCODING-010: Large message payload', async () => {
    const ctx = await setupTest('encoding-010');
    contexts.push(ctx);

    const stream = randomStreamName();
    
    // Create ~100KB of data
    const largeData: Record<string, string> = {};
    for (let i = 0; i < 1000; i++) {
      largeData[`key${i}`] = 'x'.repeat(100);
    }
    
    // Should either succeed or give clear size limit error
    try {
      const result = await ctx.client.streamWrite(stream, {
        type: 'TestEvent',
        data: largeData
      });
      expect(result.position).toBeGreaterThanOrEqual(0);
    } catch (error: any) {
      // If it fails, should be a clear error about size
      expect(error.message).toBeDefined();
    }
  });
});
